python/ray/dag/dag_node.py

```diff
@@ -82,9 +83,44 @@ def get_stable_uuid(self) -> str:
         """
         return self._stable_uuid
 
+    async def get_object_ref_from_last_execute(self, node_uuid: str):
+        """After this DAG is executed through execute(), retrieves a reference to the
+        return value of the default executor on the inputted DAGNode. The input
+        parameter should be an UUID for a method or function node.
+
+        Args:
+            node_uuid: stable uuid of a method or function DAGNode, which should be
+            part of the DAG rooted at THIS DAGNode.
+
+        Returns:
+            Object ref to the return value of the default executor node._execute_impl()
+            that was executed through calling execute() on THIS DAGNode.
+        """
+        if node_uuid not in self.cache_from_last_execute:
+            raise KeyError(
+                "Node UUID was not found in the deployment graph, unable to retrieve "
+                "object ref from last call to execute() for node."
+            )
+
+        value = self.cache_from_last_execute[node_uuid]
+        if isinstance(value, asyncio.Task):
+            return await value
+        else:
+            return value
+
     def execute(self, *args, **kwargs) -> Union[ray.ObjectRef, ray.actor.ActorHandle]:
-        """Execute this DAG using the Ray default executor."""
-        return self.apply_recursive(lambda node: node._execute_impl(*args, **kwargs))
+        """Execute this DAG using the Ray default executor.
+
+        After execution, stores the return values of the executing the default executor
+        on each node in this DAG in a cache.
+        """
+
+        def executor(node):
+            return node._execute_impl(*args, **kwargs)
+
+        result = self.apply_recursive(executor)
+        self.cache_from_last_execute = executor.cache
```

Comment:
this isn't declared in init method

Reply:
also is there a way to make this opt-in? Caching references is generally not a good idea because the adds memory pressure for users who don't need this feature. We want to make sure this feature doesn't bring extra cost to non-user.

Reply:
We can by adding a series of flags in `DAGDriver`, `RayServeDAGHandle` and `DAGNode`.